File: go/spore-core/tools/define.go
// Ergonomic custom-tool definition: DefineTool collapses the per-tool
// boilerplate of a hand-written sporecore.Tool implementation — the empty
// struct, the Name / IsSubagentTool / MayProduceLargeOutput methods, the
// Schema constructor, the input-JSON unmarshal + recoverable "invalid
// parameters" handling, and the `var _ Tool = ...` assertion — into a single
// generic adapter that returns a ready-to-register StandardTool.
//
// This is the Go analog of Rust's `tool!` macro
// (rust/crates/spore-core/src/macros.rs). The headline Rust feature is that the
// input type is the single source of truth and the JSON schema is *derived*
// from it (via schemars), so the advertised schema and the deserialization can
// never drift.
//
// Go takes the explicit-schema variant by design: the caller passes the
// parameter schema explicitly rather than pulling in a reflection-based
// JSON-schema dependency. This keeps the core module dependency-free (no
// jsonschema lib in go.mod). Reflection-based schema derivation from T is a
// possible later opt-in, but it is intentionally out of scope here.
//
// What DefineTool DOES mirror from the Rust macro:
//
//  1. It synthesizes the trivial Tool methods so the caller writes none of
//     them.
//  2. The caller provides a name, description, annotations, the explicit
//     parameter schema, and an exec function typed over the input T. The
//     adapter unmarshals the model's raw arguments into T; on failure it
//     returns a RECOVERABLE error ToolOutput whose message contains "invalid
//     parameters" (matching the Rust semantics) so a configured
//     ToolCallRepair can coerce the arguments and re-dispatch.
//  3. It returns a registrable StandardTool (Tool implementation + schema),
//     ready for HarnessBuilder.Tool(...).
//
// Example:
//
//	type calcInput struct {
//		Expression string `json:"expression"`
//	}
//	calc := tools.DefineTool(
//		"calculator",
//		"Evaluates a mathematical expression and returns the result",
//		sporecore.ToolAnnotations{},
//		json.RawMessage(`{"type":"object","properties":{"expression":{"type":"string"}},"required":["expression"]}`),
//		func(ctx context.Context, in calcInput, _ sporecore.SandboxProvider, _ *sporecore.ToolContext) sporecore.ToolOutput {
//			return sporecore.NewToolOutputSuccess("evaluated: " + in.Expression)
//		},
//	)
//	// calc plugs straight into HarnessBuilder.Tool(calc).

package tools

import (
	"context"
	"encoding/json"
	"fmt"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// ToolExecFunc is the typed execution seam DefineTool wraps. It mirrors the
// sporecore.Tool.Execute signature (context, sandbox, *ToolContext) but takes
// the already-unmarshaled, typed input T in place of the raw ToolCall — the
// adapter does the JSON decode for you. Return a ToolOutput exactly as a
// hand-written Tool.Execute would: NewToolOutputSuccess / NewToolOutputError
// (recoverable) / NewToolOutputFatal.
type ToolExecFunc[T any] func(ctx context.Context, input T, sandbox sporecore.SandboxProvider, toolCtx *sporecore.ToolContext) sporecore.ToolOutput

// DefineTool bundles a typed exec function with its metadata and explicit
// parameter schema into a registrable StandardTool, synthesizing the trivial
// Tool methods so the caller writes none of them.
//
// The model's raw tool arguments are unmarshaled into T before execFn runs. If
// they fail to decode (a missing required field, a wrong-typed field, malformed
// JSON), DefineTool short-circuits with a RECOVERABLE error ToolOutput whose
// message contains "invalid parameters" — so a configured ToolCallRepair gets a
// chance to coerce the arguments and re-dispatch, rather than halting the run.
// This matches Rust's `tool!` macro semantics.
//
// schema is the explicit JSON-Schema object advertised to the model (it must be
// a JSON object with a top-level "type", as the registry validates at
// registration time). annotations carries the behavioural flags (ReadOnly /
// Destructive / Idempotent / OpenWorld) that drive concurrency and risk.
func DefineTool[T any](
	name string,
	description string,
	annotations sporecore.ToolAnnotations,
	schema json.RawMessage,
	execFn ToolExecFunc[T],
) StandardTool {
	impl := &definedTool[T]{name: name, required: requiredFields(schema), execFn: execFn}
	return StandardTool{
		Implementation: impl,
		Schema: sporecore.RegistryToolSchema{
			Name:        name,
			Description: description,
			Parameters:  schema,
			Annotations: annotations,
		},
	}
}

// definedTool is the synthesized sporecore.Tool implementation backing every
// DefineTool result. It is unexported: callers only ever see the StandardTool.
type definedTool[T any] struct {
	name string
	// required is the schema's top-level `required` field names, extracted once
	// at construction. Go's json.Unmarshal silently leaves a missing field at
	// its zero value (unlike Rust's serde, which errors on a missing required
	// field), so the adapter checks presence explicitly to reproduce Rust's
	// "missing required field → recoverable invalid parameters" semantics.
	required []string
	execFn   ToolExecFunc[T]
}

// requiredFields extracts the top-level `required` array from a JSON-Schema
// object. A malformed or absent `required` yields nil (no presence checks),
// matching the registry's own best-effort validateInput.
func requiredFields(schema json.RawMessage) []string {
	var params struct {
		Required []string `json:"required"`
	}
	if err := json.Unmarshal(schema, &params); err != nil {
		return nil
	}
	return params.Required
}

// Name implements sporecore.Tool. It returns the registered name, which must
// match the bundled schema's Name (the registry enforces this).
func (t *definedTool[T]) Name() string { return t.name }

// IsSubagentTool implements sporecore.Tool. A DefineTool tool wraps no child
// harness — subagent tools are built through the dedicated subagent path, not
// this convenience helper.
func (t *definedTool[T]) IsSubagentTool() bool { return false }

// MayProduceLargeOutput implements sporecore.Tool. DefineTool tools are not
// assumed to produce large output; a tool that streams large results should be
// hand-written so it can opt into SandboxProvider.HandleLargeOutput.
func (t *definedTool[T]) MayProduceLargeOutput() bool { return false }

// Execute implements sporecore.Tool. It unmarshals the call's raw input into T
// and dispatches to execFn, returning a recoverable "invalid parameters" error
// ToolOutput if the decode fails.
func (t *definedTool[T]) Execute(
	ctx context.Context,
	call sporecore.ToolCall,
	sandbox sporecore.SandboxProvider,
	toolCtx *sporecore.ToolContext,
) sporecore.ToolOutput {
	// An absent/empty body is treated as `{}` so a tool with no required fields
	// still runs; a tool WITH required fields fails the presence check below.
	raw := call.Input
	if len(raw) == 0 {
		raw = json.RawMessage("{}")
	}

	// Presence check for required fields FIRST. Go's json.Unmarshal silently
	// leaves a missing field at its zero value, so without this a missing
	// required field would slip through as the zero value rather than the
	// recoverable "invalid parameters" error Rust's serde produces. Decoding
	// into a generic object also surfaces malformed JSON as a recoverable error.
	var probe map[string]json.RawMessage
	if err := json.Unmarshal(raw, &probe); err != nil {
		return sporecore.NewToolOutputError(
			fmt.Sprintf("invalid parameters for tool %q: %v", t.name, err),
		)
	}
	for _, field := range t.required {
		if _, present := probe[field]; !present {
			return sporecore.NewToolOutputError(
				fmt.Sprintf("invalid parameters for tool %q: missing required field %q", t.name, field),
			)
		}
	}

	// Now decode into the typed input. A present-but-wrong-typed field (e.g. a
	// numeric `key` where a string is expected) errors here — also recoverable.
	var input T
	if err := json.Unmarshal(raw, &input); err != nil {
		return sporecore.NewToolOutputError(
			fmt.Sprintf("invalid parameters for tool %q: %v", t.name, err),
		)
	}
	return t.execFn(ctx, input, sandbox, toolCtx)
}

// Compile-time check that the synthesized implementation satisfies Tool.
var _ sporecore.Tool = (*definedTool[struct{}])(nil)

File: python/packages/spore_tools/src/spore_tools/tools/define.py
"""Ergonomic tool definition — :func:`define_tool`.

Mirrors Rust's ``tool!`` macro (``rust/crates/spore-core/src/macros.rs``):
:func:`define_tool` collapses the boilerplate of a hand-written
:class:`~spore_core.tool_registry.Tool` implementation — name, schema, input
validation, and an async ``execute`` body — into a single call that returns a
ready-to-register :class:`~spore_tools.StandardTool`.

The headline property is the same as the Rust macro: **the input model is the
single source of truth.** The advertised JSON Schema the model sees is *derived*
from the pydantic input model via ``model.model_json_schema()`` — never
hand-written — so the schema and the validation can never drift apart.

If the model's tool arguments fail to validate against the input model, the
generated tool returns a **recoverable** :class:`~spore_core.harness.ToolOutputError`
(message ``invalid parameters for tool `{name}`: {e}``) so a configured tool-call
repair pass gets a chance to coerce the arguments and re-dispatch, rather than
halting the run.

Example::

    from pydantic import BaseModel
    from spore_core.harness import ToolOutputSuccess
    from spore_tools import define_tool

    class CalculatorInput(BaseModel):
        expression: str

    async def run(input: CalculatorInput, sandbox, ctx):
        return ToolOutputSuccess.success(f"evaluated: {input.expression}")

    calculator = define_tool(
        name="calculator",
        description="Evaluates a mathematical expression and returns the result",
        input_model=CalculatorInput,
        execute=run,
    )
    # calculator.schema.name == "calculator"
    # calculator.schema.parameters is derived from CalculatorInput

The returned :class:`~spore_tools.StandardTool` plugs straight into a builder via
:meth:`~spore_core.HarnessBuilder.tool`.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any, TypeVar

from pydantic import BaseModel, ValidationError

from spore_core.harness import (
    SandboxProvider,
    ToolOutput,
    ToolOutputError,
)
from spore_core.model import ToolCall
from spore_core.tool_registry import ToolAnnotations, ToolContext, ToolSchema

from .catalogue import StandardTool

InputT = TypeVar("InputT", bound=BaseModel)

#: An async ``execute`` body: it receives the *validated* input model plus the
#: two seams (the environment ``sandbox`` and the storage ``ctx``) and returns a
#: :class:`~spore_core.harness.ToolOutput`. This mirrors the Rust closure
#: signature ``|input, sandbox, ctx| async move { .. }``.
ExecuteFn = Callable[[InputT, SandboxProvider, ToolContext], Awaitable[ToolOutput]]


def _schema_for_model(model: type[BaseModel]) -> dict[str, Any]:
    """Derive a tool's ``parameters`` JSON Schema from its pydantic input model.

    Used by :func:`define_tool` so the advertised schema is generated from the
    exact model the tool validates. Falls back to a permissive
    ``{"type": "object"}`` if the model ever fails to produce a schema (it does
    not for any well-formed :class:`pydantic.BaseModel`)."""
    try:
        schema = model.model_json_schema()
    except Exception:  # noqa: BLE001 — never let schema derivation crash registration
        return {"type": "object"}
    # pydantic always emits ``type: object`` for a model, but guard so the
    # registry's structural check (a top-level ``type``) always passes.
    if "type" not in schema:
        schema["type"] = "object"
    return schema


class _DefinedTool:
    """Anonymous :class:`~spore_core.tool_registry.Tool` implementation backing a
    :func:`define_tool` call. Validates ``call.input`` into the input model and,
    on failure, returns a recoverable ``invalid parameters`` error; otherwise it
    forwards the validated model to the user's ``execute`` body."""

    def __init__(
        self,
        name: str,
        input_model: type[BaseModel],
        execute: ExecuteFn[Any],
        *,
        may_produce_large_output: bool,
    ) -> None:
        self._name = name
        self._input_model = input_model
        self._execute = execute
        self._may_produce_large_output = may_produce_large_output

    def name(self) -> str:
        return self._name

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return self._may_produce_large_output

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            parsed = self._input_model.model_validate(call.input)
        except ValidationError as e:
            # Recoverable so a tool-call-repair pass can coerce the arguments and
            # re-dispatch — mirrors the Rust macro's deserialize-failure branch.
            return ToolOutputError(
                message=f"invalid parameters for tool `{self._name}`: {e}",
                recoverable=True,
            )
        return await self._execute(parsed, sandbox, ctx)


def define_tool(
    *,
    name: str,
    description: str,
    input_model: type[InputT],
    execute: ExecuteFn[InputT],
    annotations: ToolAnnotations | None = None,
    may_produce_large_output: bool = False,
) -> StandardTool:
    """Define a :class:`~spore_core.tool_registry.Tool` and bundle it with its
    *derived* schema into a :class:`~spore_tools.StandardTool` in one call.

    The Python analogue of Rust's ``tool!`` macro. The advertised JSON Schema is
    derived from ``input_model`` via ``model_json_schema()`` so it can never
    drift from the model the tool actually validates.

    Args:
        name: The tool's registered name (must be non-empty).
        description: Human-readable description shown to the model.
        input_model: A :class:`pydantic.BaseModel` subclass. The tool validates
            the model's tool arguments into this and derives the advertised JSON
            Schema from it.
        execute: An async callable ``(input, sandbox, ctx) -> ToolOutput``.
            ``input`` is the validated ``input_model`` instance; ``sandbox`` is
            the :class:`~spore_core.harness.SandboxProvider` (environment seam);
            ``ctx`` is the :class:`~spore_core.tool_registry.ToolContext`
            (storage seam).
        annotations: Optional :class:`~spore_core.tool_registry.ToolAnnotations`.
            Defaults to all-``False`` (a read-write, non-idempotent tool) when
            omitted — matching the Rust macro's ``ToolAnnotations::default``.
        may_produce_large_output: Whether the tool may emit large output (drives
            large-output routing). Defaults to ``False``.

    Returns:
        A :class:`~spore_tools.StandardTool` ready for
        :meth:`~spore_core.HarnessBuilder.tool`.

    If the model's arguments fail to validate against ``input_model``, the
    generated tool returns a **recoverable**
    :class:`~spore_core.harness.ToolOutputError` whose message contains
    ``invalid parameters`` — so tool-call repair gets a chance to retry.
    """
    impl = _DefinedTool(
        name,
        input_model,
        execute,
        may_produce_large_output=may_produce_large_output,
    )
    schema = ToolSchema(
        name=name,
        description=description,
        parameters=_schema_for_model(input_model),
        annotations=annotations if annotations is not None else ToolAnnotations(),
    )
    return StandardTool(implementation=impl, schema=schema)


__all__ = ["ExecuteFn", "define_tool"]
